pkg/kubelet/cm/topologymanager/topology_manager.go

```diff
@@ -54,3 +67,217 @@ type TopologyHint struct {
+	iterate = func(i int, accum []TopologyHint) {
+		// Base case: we have looped through all providers and have a full permutation.
+		if i == len(allProviderHints) {
+			callback(accum)
+			return
+		}
+
+		// Loop through all hints for provider 'i', and recurse to build the
+		// the permutation of this hint with all hints from providers 'i++'.
+		for j := range allProviderHints[i] {
+			iterate(i+1, append(accum, allProviderHints[i][j]))
+		}
+	}
+	iterate(0, []TopologyHint{})
+}
+
+// Merge the hints from all hint providers to find the best one.
+func (m *manager) calculateAffinity(pod v1.Pod, container v1.Container) TopologyHint {
+	// Set the default hint to return from this function as an any-socket
+	// affinity with an unpreferred allocation. This will only be returned if
+	// no better hint can be found when merging hints from each hint provider.
+	defaultAffinity, _ := socketmask.NewSocketMask()
+	defaultAffinity.Fill()
+	defaultHint := TopologyHint{defaultAffinity, false}
+
+	// Loop through all hint providers and save an accumulated list of the
+	// hints returned by each hint provider. If no hints are provided, assume
+	// that provider has no preference for topology-aware allocation.
+	var allProviderHints [][]TopologyHint
+	for _, provider := range m.hintProviders {
+		// Get the TopologyHints from a provider.
+		hints := provider.GetTopologyHints(pod, container)
+
+		// If hints is nil, overwrite 'hints' with a preferred any-socket affinity.
+		if hints == nil {
+			klog.Infof("[topologymanager] Hint Provider has no preference for socket affinity")
+			affinity, _ := socketmask.NewSocketMask()
+			affinity.Fill()
+			hints = []TopologyHint{{affinity, true}}
+		}
+
+		// Accumulate the sorted hints into a [][]TopologyHint slice
+		allProviderHints = append(allProviderHints, hints)
+	}
+
+	// Iterate over all permutations of hints in 'allProviderHints'. Merge the
+	// hints in each permutation by taking the bitwise-and of their affinity masks.
+	// Return the hint with the narrowest SocketAffinity of all merged
+	// permutations that have at least one socket set. If no merged mask can be
+	// found that has at least one socket set, return the 'defaultHint'.
+	bestHint := defaultHint
+	m.iterateAllProviderTopologyHints(allProviderHints, func(permutation []TopologyHint) {
+		// Get the SocketAffinity from each hint in the permutation and see if any
+		// of them encode unpreferred allocations.
+		preferred := true
+		var socketAffinities []socketmask.SocketMask
+		for _, hint := range permutation {
+			// Only consider hints that have an actual SocketAffinity set.
+			if hint.SocketAffinity != nil {
+				if !hint.Preferred {
+					preferred = false
+				}
+				socketAffinities = append(socketAffinities, hint.SocketAffinity)
+			}
+		}
+
+		// Merge the affinities using a bitwise-and operation.
+		mergedAffinity, _ := socketmask.NewSocketMask()
+		mergedAffinity.Fill()
+		mergedAffinity.And(socketAffinities...)
```

Comment:
Can you explain why to do And and not OR? So I think you want  to create mergedAffinity which has the socketAffinities of all the hints in the  permutation. Then in Line 228 you will take the Narrower socketAffint hint anyway.
So if the device  manager hinst is [(1,1 prefered= true)]  and the cpu manager is [(0,1 prefered= true), (1,0 prefered= true), (1,1 prefered= true)] the topology manger hint will be (1,1 prefered= true).

Reply:
Hmm. I'm a bit confused by your example, because I would actually expect the correct output of the TopologyManager after the merge to be either (0,1 preferred= true) or (1,0 preferred= true), but never (1,1 preferred= true).

In one sense this is a contrived example, however, because the CPUManager should never return hints of [(0,1 preferred= true), (1,0 preferred= true), (1,1 preferred= true)]. The (1,1) case will never be a preferred allocation if it is possible to allocate all requested CPUs from either (0,1) or (1,0).

That is to say, the only time you should ever set (1,1 preferred= true) for any hint provider is if there is no other NUMA alignment possible (i.e. you only set (1,1 preferred= true) if the allocation request can only be satisfied by allocating devices from both sockets).

As such, the merge algorithm is currently designed to search for the narrowest socket mask that satisfies alignment on (at least) one of the sockets from (at least) one of the preferred socket mask hints from all providers. I.e. it treats the mask as an actual _mask_, where setting a bit in the mask indicates that is is OK to allocate a device with affinity to that socket (but it is not necessarily required).

In your example above (adjusted to account for the proper setting of the preferred field for the last hing in the CPUManager), this would generate the following merged hints:
```
[(1,1, preferred=true), (0,1, preferred=true)] = (0,1, preferred=true)
[(1,1, preferred=true), (1,0, preferred=true)] = (1,0, preferred=true)
[(1,1, preferred=true), (1,1, preferred=false)] = (1,1, preferred=false)
```
With (0,1, preferred=true) winning out in the end.

It sounds like you are promoting a different algorithm, however. One that guarantees allocation of devices according to exact matches in socket affinities from all hint providers. I.e one that would result in the following (I drop the preferred flag here because its existence would be meaningless in this setup):
```
[(1,1), (0,1)] = (0,0)
[(1,1), (1,0)] = (0,0)
[(1,1), (1,1)] = (1,1)
```
With (1,1) winning out, and the devicemanager and CPUManager now being told to explicitly allocate devices from *both* sockets (rather than interpreting the socket mask as an "OK" to allocate from one, the other, or both).

Reply:
yes you are absolutely right, that was my interpretation of strict mode. (and why I didn't understand the need for preferred flag:) )

Reply:
I think the original comment was resolved, and the discussion on the strict poilcy interpretation can be moved to Kevin's google doc. So I am going to resolve this. Thanks for the discussion!